### ISSUE TITLE
Support client side load balancing

### ISSUE THREAD

**Issue by npepinpe** [camunda/camunda#9870]
**Is your feature request related to a problem? Please describe.**

While most users will typically put a load balancer in front of their gateways, for testing and development purposes it's often the case that you simply deploy multiple instances without anything in front.

**Describe the solution you'd like**

As gRPC Java already support load balancing, I would love to be able to simply provide my set of instances and let the client do the load balancing, without having to deploy any kind of balancer/proxy.

**Describe alternatives you've considered**

Possibly we could offer an escape hatch in the client builder which allows one to customize the `ManagedChannel`'s builder, without having to support explicit client side load balancing.

**Additional context**

This is a pure feature request from side after playing with deploying clusters in Docker Compose and Testcontainers, where I then have to add wrapping code in my client to take care of switching instances (or create new clients).

I understand this is probably not a big issue for production use cases, of course :slightly_smiling_face: 

### Acceptance Criteria:

- [x] a client side load balancing feature is added following the approach of https://github.com/camunda/camunda/commit/b8697cb3a9cf396032100616bc7f16236d39740d for gRPC or equivalent
- [x] for the apache httpClient created in io.camunda.client.impl.http.HttpClientFactory a `RoundRobinDnsResolver` is implemented that rotates resolved addresses for a host, effectively load balancing across all resolved addresses
- [x] a config flag exists to opt-in to make use of the load balancing feature
- [x] an assessment and if possible Integration tests of scenarios/setups that could break due to this change: at least `headless k8s services (load balancer w/ single ip)`, `plain ip configured as grpcAddress`

**Comment by Zelldon** [comment]
This would mean that we have to implement for all others clients as well right? 

Regarding the load-balancing, what do you mean just sending commands to different gateways?  OR sending commands direct to certain partitions?

BTW Might also be related to #11310

**Comment by npepinpe** [comment]
Send requests to different gateways. Grpc already supports out of the box client side load balancing, so it would be just exposing or hooking it up. It's for deployments where you don't have a proper/good load balancer in front. It's not super urgent or anything, as I expect most users with multiple gateways would have a load balancer in front.

**Comment by shahamit** [comment]
@npepinpe - we are facing this issue. We want to try the workaround of grpc based client side load balancing. We are using [this](https://docs.camunda.io/docs/apis-tools/java-client/) `zeebe-client-java` library to connect to the zeebe cluster. I understand that we can enable the grpc client side load balancing through the `io.grpc.ManagedChannel` interface. 

Could you please guide me how to enable it through the `zeebe-client-java` library?

Thanks.

**Comment by ChrisKujawa** [comment]
@megglos I put it on your teams board, because it is client and camunda user experience related. We [showed](https://camunda.slack.com/archives/C0807665N8G/p1764267579243299?thread_ts=1764249628.788339&cid=C0807665N8G) that with having this (see [POC](https://github.com/camunda/camunda/pull/38916)) we can better distribute the load, reaching higher throughput (with keeping latency).


The missing part of the POC is that we need configuration for the client, fex. toggle to turn on and off. Could your team look into this?

**Comment by ChrisKujawa** [comment]
@megglos i would like to discuss this whether we can get support for this - the fix is rather small https://github.com/camunda/camunda/pull/44770/changes/b8697cb3a9cf396032100616bc7f16236d39740d but it think we need to make sure that we have a feature flag etc

**Comment by megglos** [comment]
> [@megglos](https://github.com/megglos) i would like to discuss this whether we can get support for this - the fix is rather small [b8697cb](https://github.com/camunda/camunda/commit/b8697cb3a9cf396032100616bc7f16236d39740d) but it think we need to make sure that we have a feature flag etc

What's the risk that comes with it that makes you want to bind it to a feature flag? From my understanding this will not break any existing setups but behavior just changes that the client will rotate across the resolved IPs? @entangled90 @ChrisKujawa 

Or is it more like a precaution measure? I tend to prefer it as default from my current understanding

**Comment by entangled90** [comment]
I think we need to do some tests there to understand what might be the issue. A couple of scenario from the top of my mind:
- k8s services is not headless (load balancer w/ single ip)
- static ip (e.g. `218.199.28.39`)
Maybe it does not work in one of those two cases, then enabling by default it's a breaking change

**Comment by megglos** [comment]
@ChrisKujawa @entangled90 I wonder, we probably want the same behavior also for the apache http client used for the REST API, right? Or would a gRPC specific config flag be suffice?

**Comment by entangled90** [comment]
Not sure if it's really needed, we should run some tests first maybe? How many connection does the client open? do we have a keep-alive? I expect that after some time those connections are closed anyway and reopened, which will balance the load over time.

**Comment by megglos** [comment]
> Not sure if it's really needed, we should run some tests first maybe? How many connection does the client open? do we have a keep-alive? I expect that after some time those connections are closed anyway and reopened, which will balance the load over time.

As far as I understand https://github.com/apache/httpcomponents-client/blob/master/httpclient5/src/main/java/org/apache/hc/client5/http/impl/io/DefaultHttpClientConnectionOperator.java#L204 the apache client will just always use the first ip it can connect to if there are multiple resolved. So it it will always connect to the same ip, unless connection fails. So even with short lived requests they may all end up with the same gateway.

**Comment by entangled90** [comment]
Yeah so the problem boils down to the fact that grpc needs a headless k8s service, while http requires a load balancer. I'd say we move towards the headless service for both and do something similar for http?

**Comment by ChrisKujawa** [comment]
But i think this only works for SM and having control over the headless service. If you use SaaS or an ingress the story is different i think

**Comment by entangled90** [comment]
For Saas I think nothing is required from the client perspective, we just need to configure NGINX to use the headless service instead, see [grpc_pass](https://nginx.org/en/docs/http/ngx_http_grpc_module.html#grpc_pass).

A client opens a single TCP stream to NGINX, which then dispatches each rpc to different servers:
> If a domain name resolves to several addresses, all of them will be used in a round-robin fashion. In addition, an address can be specified as a [server group](https://nginx.org/en/docs/http/ngx_http_upstream_module.html).

**Comment by megglos** [comment]
https://github.com/camunda/camunda/pull/48533#issuecomment-4074609085
